### PR TITLE
Sanitize use of external macros

### LINKED ISSUE #460: libcu++ should not define __host__/__device__/__forceinline__ (or any other runtime macros)
In our `__config` we have the following:

https://github.com/NVIDIA/libcudacxx/blob/b8b37d6784188cf4e64eb2f04d89797e8f77bea1/include/cuda/std/detail/__config#L14-L39

this will define things like `__host__` to be empty when we detect that we're not compiling with nvcc in order to silence any errors about unrecognized symbols that would occur from host compilers that don't understand what `__host__` is. 

That is a reasonable thing in many cases, but there are legitimate use cases where this can break. For example, the following file compiled with a host compiler will fail to compile:
```
#include <cuda/std/type_traits>
#include <cuda_runtime_api.h>
```

This will fail due to a redefinition error of `__host__` and friends because `cuda/std/type_traits` will detect we're not compiling with nvcc and define `__host__` to empty, but then `cuda_runtime_api.h` will _also_ define `__host__`, causing the redefinition error. 

One could reasonably argue that the runtime headers should better guard against this situation by undefining `__host__` before defining it. However, it is most expedient for us to fix this in libcu++. 

The only robust solution I can think of is to avoid ever defining `__host__`/`__device__`/`__forceinline__` ourselves and instead use our own macros like:
```
#if defined(__CUDACC__)
  #define _LIBCUDACXX_HOST __host__
#else 
  #define _LIBCUDACXX_HOST
#endif
```

And use those in place of any where we'd use `__host__` (or other macros). 

I'd expect we should rarely be using host/device annotations directly anyways. Those should come from [`__LIBCUDACXX_INLINE_VISIBILITY`](https://github.com/NVIDIA/libcudacxx/blob/b8b37d6784188cf4e64eb2f04d89797e8f77bea1/include/cuda/std/detail/libcxx/include/__config#L1349-L1350) or just `_LIBCUDACXX_EXECUTION_SPACE_SPECIFIER`.

### PR DESCRIPTION
**Pull request by miscco**
Currently we define certain external macros that are used in product code.

However, this might clash with external definitions and create include order issues.

To alleviate this, define internal `_LIBCUDACXX_MEOW` macros that cannot clash with the outside world.

We also defined either `__float16` or `_Float16` despite them not being used internally at all. We should not do this and this has lead to compile issues on at least GCC. So drop those definitions. 

Addresses nvbug4139266 and nvbug4123393
Fixes #460

### CODE REVIEW

File: include/cuda/std/detail/libcxx/include/__config [hunk 1]
```diff
@@ -1365,13 +1365,23 @@ typedef __char32_t char32_t;
 #  define _LIBCUDACXX_HIDE_FROM_ABI_AFTER_V1 _LIBCUDACXX_HIDE_FROM_ABI
 #endif
 
+#ifdef __CUDACC__
+#define _LIBCUDACXX_HOST __host__
+#define _LIBCUDACXX_DEVICE __device__
+#define _LIBCUDACXX_HOST_DEVICE __host__ __device__
+#define _LIBCUDACXX_FORCE_INLINE __forceinline__
+#else // ^^^ __cuda_std__ ^^^ / vvv !__cuda_std__
```
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/__config:1373]:
```suggestion
#else // ^^^__CUDACC__ / vvv !__CUDACC__
```

File: include/cuda/std/detail/libcxx/include/__config [hunk 2]
```diff
@@ -1365,13 +1365,23 @@ typedef __char32_t char32_t;
...
+#define _LIBCUDACXX_HOST
+#define _LIBCUDACXX_DEVICE
+#define _LIBCUDACXX_HOST_DEVICE
+#define _LIBCUDACXX_FORCE_INLINE
+#endif // !__cuda_std__
```
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/__config:1378]:
```suggestion
#endif // !__CUDACC__
```

File: include/cuda/std/detail/libcxx/include/__iterator/iterator_traits.h [hunk 3]
```diff
@@ -40,9 +40,9 @@ struct __has_iterator_typedefs
 private:
     struct __two {char __lx; char __lxx;};
     template <class _Up>
-    _LIBCUDACXX_EXECUTION_SPACE_SPECIFIER _LIBCUDACXX_INLINE_VISIBILITY static __two __test(...);
+    _LIBCUDACXX_HOST_DEVICE _LIBCUDACXX_INLINE_VISIBILITY static __two __test(...);
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply include/cuda/std/detail/libcxx/include/__iterator/iterator_traits.h:43]: I'm surprised nvcc doesn't warn on redundantly specifying `__host__ __device__` annotations: https://godbolt.org/z/vqErcGvcY

[the remark this answers is not in the document]
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/__iterator/iterator_traits.h:43]: Yeah, removed those

File: include/cuda/std/detail/libcxx/include/__config [hunk 4]
```diff
@@ -1365,13 +1365,23 @@ typedef __char32_t char32_t;
...
+
 // Just so we can migrate to the new macros gradually.
 
 #ifdef __cuda_std__
-#  define _LIBCUDACXX_EXECUTION_SPACE_SPECIFIER __host__ __device__
-#  define _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_EXECUTION_SPACE_SPECIFIER
+#  define _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_HOST_DEVICE
```
[the remark this answers is not in the document]
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/__config:1383]: The precedent from upstream is to only ever use `_LIBCUDACXX_HIDE_FROM_ABI` so I think that should be the route to go

**Review by jrhemstad — APPROVED** [review]

**Review by wmaxey — APPROVED** [review]

**Pull request merged**